lib/middleware/files.js

```diff
@@ -8,44 +8,107 @@
 
-    if (hasTrailingSlash && configTrailingSlash === false) {
-      return res.superstatic.handle({redirect: withoutTrailingSlash});
-    } else if (!hasTrailingSlash && configTrailingSlash === true) {
-      return res.superstatic.handle({redirect: withTrailingSlash});
-    }
+      // We want to check for some other files, namely an `index.html` if this were a directory.
+      var pathAsDirectoryWithIndex = pathutils.asDirectoryIndex(pathutils.addTrailingSlash(pathname));
+      return res.superstatic.provider(req, pathAsDirectoryWithIndex).then(function(pathAsDirectoryWithIndexResult) {
+        // If an exact file wins now, we know that this path leads us to a directory.
+        if (pathAsDirectoryWithIndexResult) {
+          if (trailingSlashBehavior === undefined) {
+            if (!hasTrailingSlash) {
+              if (!cleanUrlRules) {
+                return res.superstatic.handle({redirect: pathutils.addTrailingSlash(pathname) + search});
+              }
+            }
+          } else if (trailingSlashBehavior === false) {
+            if (hasTrailingSlash) {
+              return res.superstatic.handle({redirect: pathutils.removeTrailingSlash(pathname) + search});
+            }
+          } else if (trailingSlashBehavior === true) {
+            if (!hasTrailingSlash) {
+              return res.superstatic.handle({redirect: pathutils.addTrailingSlash(pathname) + search});
+            }
+          }
+          // If we haven't returned yet, our path is "correct" and we should be serving a file, not redirecting.
+          return res.superstatic.handleFileStream({file: pathAsDirectoryWithIndex}, pathAsDirectoryWithIndexResult);
+        }
 
-    var fetchPath = hasTrailingSlash ? indexPathname : pathname;
-    var fallbackPath = hasTrailingSlash ? pathname : indexPathname;
-    var noresultRedirect = hasTrailingSlash ? withoutTrailingSlash : withTrailingSlash;
+        // Let's check on the clean URLs property.
+        // We want to know if a specific mutation of the path exists.
+        if (cleanUrlRules) {
+          var appendedPath = pathname;
+          if (hasTrailingSlash) {
+            if (trailingSlashBehavior !== undefined) {
+              // We want to remove the trailing slash and see if a file exists with an .html attached.
+              appendedPath = pathutils.removeTrailingString(pathname, '/') + '.html';
+            }
+          } else {
+            // Let's see if our path is a simple clean URL missing a .HTML5
+            appendedPath += '.html';
+          }
 
-    return res.superstatic.provider(req, fetchPath).then(function(result) {
-      if (result) {
-        return res.superstatic.handleFileStream({file: fetchPath}, result);
-      }
+          return res.superstatic.provider(req, appendedPath).then(function(appendedPathResult) {
+            if (appendedPathResult) {
+              // Okay, back to trailing slash behavior
+              if (trailingSlashBehavior === false) {
```

Comment:
Would again suggest unnesting these for clarity.

Reply:
Done.